### DISCUSSION TITLE
ARC-0044: Building an EpochProgram of Similar Computational Loads

### DISCUSSION THREAD

**Discussion by OzielLa** [ProvableHQ/ARCs#78] [ARCs]
---
arc: ARC-0044
title: Building an EpochProgram of Similar Computational Loads
authors: OzielLa
topic: Protocol
status: Draft
created: 10/17/2024
---

## Abstract

This proposal aims to address the issue of  1 to 4 times difference in puzzle- solving workload across different epochs by limiting the number of `num_private` instead of `opcode` . When the overall computational power of provers across the network stabilizes and significant fluctuations in the Proof Target occur, Puzzle rewards may vary by 1 to 4 times. These frequent fluctuations impact the estimated output model of the project.
<img width="1362" alt="image" src="https://github.com/user-attachments/assets/d340e170-f5b2-4b92-82e1-0eeded230424">

## Background

After a period of operation and technical analysis, it has been observed that the workload of puzzles has a significant correlation with the quantity of `num_private` generated during program execution.

## Proposal

This proposal advises to replace the current limit on `opcode` with that on the quantity of `num_private`.

1. Evaluate a reasonable number of `leaves` as `MAX_LEAVES_COUNT`.
2. Build an `InstructionSet` by adding a data relationship between puzzle workloads and  the quantity of `num_private`.
3. Enhance the building process of `instruction_string`.

## Technical Considerations

1. When determining `MAX_LEAVES_COUNT`, it needs to ensure a sufficient number of Operation Instructions to guarantee sufficient randomness in the EpochProgram.
2. After introducing `num_private` into `InstructionSet`, how to reasonably  adjust the original weight values needs to be considered .

## Benefits

1. Proof Target adjustment can align with the computational power of provers.
2. No unusual fluctuations in Puzzle Rewards and Stake Rewards.

## Risks and Mitigations

TBD

## Conclusion

By building an EpochProgram of similar computational loads, that can facilitate the assessment of current network computational power change and ensure fairer returns for provers.

**Comment by xiongqiumin** [comment]
May I ask if there is sample code on how to achieve this goal

**Comment by LoiusZK** [comment]
I would like to ask how this proposal is progressing? It seems to be stagnant. I feel that this proposal is a good improvement on the current mainnet algorithm.
